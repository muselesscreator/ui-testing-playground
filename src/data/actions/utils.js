import { createAction } from '@reduxjs/toolkit';

const createActionFactory = (dataKey) => (actionKey, ...args) => (
  createAction(`${dataKey}/${actionKey}`, ...args)
);

export {
  // eslint-disable-next-line import/prefer-default-export
  createActionFactory,
};
