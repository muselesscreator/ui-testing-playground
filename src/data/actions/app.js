import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'app';
const createAction = createActionFactory(dataKey);

/**
 * Add a new roll string to the rolls list
 * @param {string} roll - new roll string
 */
export const addRoll = createAction('add_roll');

/**
 * Clears all rolls from the rolls list
 */
export const clearRolls = createAction('clear_rolls');

export default StrictDict({
  addRoll,
  clearRolls,
});
