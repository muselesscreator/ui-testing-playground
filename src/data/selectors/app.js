import { createSelector } from 'reselect';

import { StrictDict } from 'utils';

import * as module from './app';

export const appSelector = (state) => state.app;

export const simpleSelectors = {
  rolls: createSelector([module.appSelector], (app) => app.rolls),
  lastRollsByType: createSelector(
    [module.appSelector],
    (app) => app.lastRollsByType,
  ),
};

export const rollSummaries = createSelector(
  [module.simpleSelectors.rolls],
  (rolls) => rolls.map(
    ({ rollIndex, type, value }) => ({
      rollIndex,
      summary: `You rolled ${value} on a ${type}`,
    }),
  ),
);

export const lastRollByType = (state, { type }) => (
  module.simpleSelectors.lastRollsByType(state)[type]
);

export default StrictDict({
  ...simpleSelectors,
  lastRollByType,
  rollSummaries,
});
