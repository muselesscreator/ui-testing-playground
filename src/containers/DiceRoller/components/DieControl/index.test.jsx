import React from 'react';
import { shallow } from 'enzyme';

import { Image } from '@edx/paragon';

import {
  diceSides,
  diceTypes,
  diceURIs,
} from 'constants/dice';

import selectors from 'data/selectors';
import actions from 'data/actions';

import { DieControl, mapStateToProps, mapDispatchToProps } from '.';

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      lastRollByType: jest.fn((...args) => ({ lastRollByType: args })),
    },
  },
}));
jest.mock('@edx/paragon', () => ({
  Button: () => 'Button',
  Image: () => 'Image',
}));

let el;

describe('DieControl component', () => {
  const props = {
    lastRoll: 2,
    type: diceTypes.d4,
    addRoll: jest.fn().mockName('this.props.addRoll'),
  };
  const newRoll = 0.7824;
  test('snapshot', () => {
    el = shallow(<DieControl {...props} />);
    el.instance().roll = jest.fn().mockName('this.roll');
    expect(el.instance().render()).toMatchSnapshot();
  });
  describe('component', () => {
    beforeEach(() => {
      el = shallow(<DieControl {...props} />);
    });
    test('loads image src from diceURIs based on type', () => {
      expect(el.find(Image).props().src).toEqual(diceURIs[props.type]);
    });
    describe('roll callback', () => {
      beforeEach(() => {
        jest.spyOn(global.Math, 'random').mockReturnValue(newRoll);
      });
      afterEach(() => {
        jest.spyOn(global.Math, 'random').mockRestore();
      });
      it('calls addRoll prop with type and a new roll based on number of sides', () => {
        el.instance().roll();
        expect(props.addRoll).toHaveBeenCalledWith({
          type: props.type,
          value: Math.floor(newRoll * diceSides[props.type] + 1),
        });
      });
    });
  });

  describe('mapStateToProps', () => {
    const testState = { a: 'fake state' };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState, { type: props.type });
    });
    test('lastRoll is loaded from selectors.app.lastRollByType, with ownProps.type', () => {
      expect(mapped.lastRoll).toEqual(
        selectors.app.lastRollByType(testState, { type: props.type }),
      );
    });
  });

  describe('mapDispatchToProps', () => {
    test('maps actions.app.addRoll to addRoll prop', () => {
      expect(mapDispatchToProps.addRoll).toEqual(actions.app.addRoll);
    });
  });
});
