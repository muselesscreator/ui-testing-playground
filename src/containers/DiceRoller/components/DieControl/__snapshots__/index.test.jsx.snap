// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DieControl component snapshot 1`] = `
<div
  className="die-control"
>
  <Image
    alt="dice"
    className="die-img"
    src="https://s3.amazonaws.com/files.thegamecrafter.com/23ab580e2280cd20f2ef36260eb2b8902132e5f7"
  />
  <br />
  <div
    className="last-value"
  >
    Last Roll: 2
  </div>
  <Button
    className="roll-btn"
    onClick={[MockFunction this.roll]}
  >
    Roll 1
    d4
  </Button>
</div>
`;
